File: C++/Easy/Accepted/find-the-index-of-the-first-occurrence-in-a-string_8.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        
        if (needle.empty()) return 0;
        if (haystack.empty()) return -1;
        int i = 0;
        int j = 0;
        int m = haystack.size();
        int n = needle.size();
        while (i < m && j < n) {
            if (haystack[i] == needle[j]) {
                i++;
                j++;
            } else {
                i = i - j + 1;
                j = 0;
            }
        }
        if (j == n) return i - j;
        return -1;
    }
};

File: Rust/Easy/Accepted/find-the-index-of-the-first-occurrence-in-a-string_8.rs
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        
        if needle.len() == 0 {
            return 0;
        }
        let mut i = 0;
        let mut j = 0;
        let mut k = 0;
        let mut index = -1;
        let haystack = haystack.as_bytes();
        let needle = needle.as_bytes();
        while i < haystack.len() {
            if haystack[i] == needle[j] {
                if index == -1 {
                    index = i as i32;
                }
                j += 1;
                if j == needle.len() {
                    return index;
                }
            } else {
                if index != -1 {
                    i = k;
                    j = 0;
                    index = -1;
                }
                k += 1;
            }
            i += 1;
        }
        return -1;
    }
}